### ISSUE TITLE
Validation failure 

### ISSUE THREAD

**Issue by benjaminjellison** [GaloisInc/cryptol#494]
I wrote this code (2.5, 942173b)  to determine if one sequence of 32 bit numbers is greater than another:

```
gt_words : {a} (fin a) => [a][32] -> [a][32] -> Bit
gt_words l0 l1 = ss!0
  where
   ss = [ False ] # [ s \/ (i > j) | i <- l0 | j <- l1 | s <- ss ]
```
Seems straightforward enough, but I get this validation error:
```
[error] at ./compare.cry:5:1--7:67:
  Failed to validate user-specified signature.
    In the definition of 'Main::gt_words', at ./compare.cry:5:1--5:9:
      for any type a
        fin a
      =>
      ?o14 == min a (min a (1 + ?o14))
        arising from
        list comprehension
        at ./compare.cry:7:21--7:67
      fin ?o14
        arising from
        use of expression (!)
        at ./compare.cry:5:18--5:22
  where
  ?o14 is length of list comprehension
```
Here ```?o14``` is the size of ```ss```.  With @jpziegler's help we found that this code type checks fine:
```
gt_words : {a} (fin a) => [a][32] -> [a][32] -> Bit
gt_words l0 l1 = ss!0
  where
   ss : [a+1]Bit
   ss = [ False ] # [ s \/ (i > j) | i <- l0 | j <- l1 | s <- ss ]
```
In addition, if you load just the original function definition without the type signature, it type checks fine, but has a very odd type:
```
Main> :t gt_words 
gt_words : {a, b, c, d} (Cmp c, a == min b (min d (1 + a)),
                         fin a) => [b]c -> [d]c -> Bit

```
In particular, the ```a == min b (min d (1 + a))``` seems very problematic.  If I add that type signature and try to load it in, I get an error very similar to the first error.

**Comment by yav** [comment]
In your first example, it looks like `?o14` is not the size of `ss` but rather the second list in the definition of `ss` (the one after the `#`).    Cryptol should be able to figure out the the only possible solution for the equation is when `?o14 == a`, I'll see if I can add some rules to teach it that.

What do you think is problematic about the `a == min b (min d (1 + a))` constraint?

**Comment by benjaminjellison** [comment]
Problematic is maybe not the right word.  But the issues are:

1) including it as a type constraint fails the type check;
2) ```a == min b (min d (1 + a)``` is really hard to parse (after some thought, I determined that it is saying ```a==b \/ a==d``` without using ```\/```);
3) it's not clear what ```a``` represents, and we are left guessing what it must be.  Like in your first comment, it appears ```a``` is the size of ```ss``` minus the first element.

I was surprised that the function uses a rather simple fold but the type checker had an issue with it.  Failing that, it took some digging to realize that the problem was it couldn't determine the size of the fold.

**Comment by benjaminjellison** [comment]
I just encountered a similar problem with one of Galois' files from the saw-script directory, in saw/examples/ecdsa/cryptol-spec:

```
[cryptol-spec]$ cryptol ref_ec_mul.cry 
┏━╸┏━┓╻ ╻┏━┓╺┳╸┏━┓╻  
┃  ┣┳┛┗┳┛┣━┛ ┃ ┃ ┃┃  
┗━╸╹┗╸ ╹ ╹   ╹ ┗━┛┗━╸
version 2.5.0 (942173b)

Loading module Cryptol
Loading module Set
Loading module Morphism
Loading module Field
Loading module ec_point_ops
Loading module ref_ec_mul
[error] at ./ref_ec_mul.cry:7:1--21:15:
  Failed to validate user-specified signature.
    In the definition of 'ref_ec_mul::ec_mul', at ./ref_ec_mul.cry:7:1--7:7:
      for any type fv, gv
        fin gv
      =>
      ?w39 == min (2 + gv) (min (2 + gv) (1 + ?w39))
        arising from
        list comprehension
        at ./ref_ec_mul.cry:12:14--21:15
      fin ?w39
        arising from
        use of expression (Cryptol::!)
        at ./ref_ec_mul.cry:7:20--7:25
  where
  ?w39 is length of list comprehension
Loading module Cryptol
```

The offending code is here:
```
/* Multiplies a large word scalar by the an affine point. */
ec_mul : {fv, gv}  (fin gv) => (PointOps fv [gv], [gv], AffinePoint fv) -> JacobianPoint fv
ec_mul (m, d, s) = res!0
  where
    k = d # [False, False]
    h = k + (k + k)
    res = [(ec_zero_point(m.field))]
          #  [ if hi && ~ki then
                 m.add(m.double(r), s)
               else if ~hi && ki then
                 m.sub(m.double(r), s)
               else
                 m.double(r)
             | hi <- h
             | ki <- k
             | r  <- res 
             ]
```

Again, it cannot figure out the size of res (minus the first element).

If I change the code to this, it loads fine:
```
/* Multiplies a large word scalar by the an affine point. */
ec_mul : {fv, gv}  (fin gv) => (PointOps fv [gv], [gv], AffinePoint fv) -> JacobianPoint fv
ec_mul (m, d, s) = res!0
  where
    k = d # [False, False]
    h = k + (k + k)
    res : [gv+3](JacobianPoint fv)
    res = [(ec_zero_point(m.field))]
          #  [ if hi && ~ki then
                 m.add(m.double(r), s)
               else if ~hi && ki then
                 m.sub(m.double(r), s)
               else
                 m.double(r)
             | hi <- h
             | ki <- k
             | r  <- res 
             ]
```

**Comment by brianhuffman** [comment]
I think what we need here is a rule that will simplify a constraint like `a == min b (min d (1 + a))` to `a == min b d`.

More generally, whenever we have a constraint like `a == min [n1,n2..nk]` (commonly generated by recursive list comprehensions) then we should be able to filter out any `ni` that is provably strictly greater than `a`.
